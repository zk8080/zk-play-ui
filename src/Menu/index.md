<!--
 * @Author: your name
 * @Date: 2021-08-15 17:34:05
 * @LastEditTime: 2021-08-15 17:37:25
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: /zk-play-ui/src/Menu/index.md
-->

## Menu

Demo:

```tsx
import React from 'react';
import { Menu } from 'zk-play-ui';

export default (props) => (
  <Menu
    defaultIndex="0"
    className="test"
    defaultOpenSubMenus={['4']}
    mode="vertical"
  >
    <Menu.Item>active</Menu.Item>
    <Menu.Item disabled>disabled</Menu.Item>
    <Menu.Item className="test-xyz">xyz</Menu.Item>
    <Menu.SubMenuItem title="dropdown">
      <Menu.Item>drop1</Menu.Item>
    </Menu.SubMenuItem>
    <Menu.SubMenuItem title="opened">
      <Menu.Item>opened1</Menu.Item>
    </Menu.SubMenuItem>
  </Menu>
);
```
