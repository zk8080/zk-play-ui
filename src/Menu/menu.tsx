import React, { createContext, FC, useState } from 'react';
import classNames from 'classnames';
import MenuItem, { MenuItemProps } from './menuItem';
import SubMenuItem, { SubMenuItemProps } from './subMenuItem';

type MenuMode = 'horizontal' | 'vertical';
type SelectCallback = (selectedIndex: string) => void;

export interface MenuProps {
  defaultIndex?: string;
  className?: string;
  mode?: MenuMode;
  style?: React.CSSProperties;
  onSelect?: SelectCallback;
  defaultOpenSubMenus?: string[];
}

interface IMenuContext {
  index: string;
  onSelect?: SelectCallback;
  mode?: MenuMode;
  defaultOpenSubMenus?: string[];
}

export const MenuContext = createContext<IMenuContext>({ index: '0' });

const Menu: React.FC<MenuProps> = (props) => {
  const {
    className,
    mode,
    defaultIndex,
    style,
    onSelect,
    children,
    defaultOpenSubMenus,
  } = props;
  const [currentActive, setActive] = useState(defaultIndex);
  const classes = classNames('viking-menu', className, {
    'menu-vertical': mode === 'vertical',
    'menu-horizontal': mode !== 'vertical',
  });

  const handleClick = (index: string) => {
    setActive(index);
    if (onSelect) {
      onSelect(index);
    }
  };
  const passedContext: IMenuContext = {
    index: currentActive ? currentActive : '0',
    onSelect: handleClick,
    mode,
    defaultOpenSubMenus,
  };

  const renderChildren = () => {
    return React.Children.map(children, (child, index) => {
      const childElement =
        child as React.FunctionComponentElement<MenuItemProps>;
      const { displayName } = childElement.type;
      if (displayName === 'MenuItem' || displayName === 'SubMenuItem') {
        return React.cloneElement(childElement, {
          index: `${index}`,
        });
      } else {
        console.error(
          'Warning: Menu has a child which is not a MenuItem component',
        );
      }
    });
  };

  return (
    <ul className={classes} style={style} data-testid="test-menu">
      <MenuContext.Provider value={passedContext}>
        {renderChildren()}
      </MenuContext.Provider>
    </ul>
  );
};

Menu.defaultProps = {
  defaultIndex: '0',
  mode: 'horizontal',
  defaultOpenSubMenus: [],
};

type IMenuComponent = FC<MenuProps> & {
  Item: FC<MenuItemProps>;
  SubMenuItem: FC<SubMenuItemProps>;
};

const RootMenu = Menu as IMenuComponent;
RootMenu.SubMenuItem = SubMenuItem;
RootMenu.Item = MenuItem;

export default RootMenu;
