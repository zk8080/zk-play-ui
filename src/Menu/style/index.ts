/*
 * @Author: your name
 * @Date: 2021-08-22 16:35:59
 * @LastEditTime: 2021-08-22 16:42:06
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: /zk-play-ui/src/Menu/style/index.ts
 */
import './index.less';
