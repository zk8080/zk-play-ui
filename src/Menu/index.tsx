/*
 * @Author: your name
 * @Date: 2021-08-22 16:42:21
 * @LastEditTime: 2021-08-22 16:42:39
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: /zk-play-ui/src/Menu/index.tsx
 */
import Menu from './menu';

export default Menu;
