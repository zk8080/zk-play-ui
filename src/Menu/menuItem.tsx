/*
 * @Author: your name
 * @Date: 2021-01-31 15:25:29
 * @LastEditTime: 2021-08-24 21:59:38
 * @LastEditors: your name
 * @Description: In User Settings Edit
 * @FilePath: /zk-play-ui/src/Menu/menuItem.tsx
 */
import React, { useContext } from 'react';
import classNames from 'classnames';
import { MenuContext } from './menu';

export interface MenuItemProps {
  index?: string;
  disabled?: boolean;
  className?: string;
  style?: React.CSSProperties;
}

const MenuItem: React.FC<MenuItemProps> = (props) => {
  const { children, className, index, disabled, style } = props;
  const context = useContext(MenuContext);
  const classes = classNames('menu-item', className, {
    'is-disabled': disabled,
    'is-active': context.index === index,
  });
  const handleClick = () => {
    if (context.onSelect && !disabled && typeof index === 'string') {
      context.onSelect(index);
    }
  };
  return (
    <li className={classes} style={style} onClick={handleClick}>
      {children}
    </li>
  );
};

MenuItem.displayName = 'MenuItem';
export default MenuItem;
