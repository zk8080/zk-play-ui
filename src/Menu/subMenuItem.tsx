/*
 * @Author: your name
 * @Date: 2021-03-01 22:32:11
 * @LastEditTime: 2021-08-15 17:38:32
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: /vikingship/src/components/Menu/subMenuItem.tsx
 */
import React, {
  FC,
  useContext,
  FunctionComponentElement,
  useState,
} from 'react';
import classNames from 'classnames';
import { MenuContext } from './menu';
import { MenuItemProps } from './menuItem';
import Icon from '../Icon/index';
import Transition from '../Transition/index';

export interface SubMenuItemProps {
  index?: string;
  title: string;
  className?: string;
}

const SubMenuItem: FC<SubMenuItemProps> = (props) => {
  const context = useContext(MenuContext);
  const { mode, defaultOpenSubMenus } = context;
  const defaultOpens = defaultOpenSubMenus as Array<string>;
  const { index, title, className, children } = props;
  const isOpend =
    index && mode === 'vertical' ? defaultOpens.includes(index) : false;
  const [menuOpen, setMenuOpen] = useState(isOpend);
  const classes = classNames('menu-item submenu-item', className, {
    'is-active': context.index === index,
    'is-opened': menuOpen,
    'is-vertical': mode === 'vertical',
  });

  const handleClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setMenuOpen(!menuOpen);
  };
  let timer: any;
  const handleMouse = (e: React.MouseEvent, toggle: boolean) => {
    clearTimeout(timer);
    e.preventDefault();
    timer = setTimeout(() => {
      setMenuOpen(toggle);
    }, 300);
  };

  const clickEvents =
    mode === 'vertical'
      ? {
          onClick: handleClick,
        }
      : {};
  const hoverEvents =
    mode !== 'vertical'
      ? {
          onMouseEnter: (e: React.MouseEvent) => {
            handleMouse(e, true);
          },
          onMouseLeave: (e: React.MouseEvent) => {
            handleMouse(e, false);
          },
        }
      : {};

  const renderClidren = () => {
    const subMenuClasses = classNames('viking-submenu', {
      'menu-opened': menuOpen,
    });
    const childrenComponent = React.Children.map(children, (child, i) => {
      const childElement = child as FunctionComponentElement<MenuItemProps>;
      if (childElement.type.displayName === 'MenuItem') {
        return React.cloneElement(childElement, {
          index: `${index}-${i}`,
        });
      } else {
        console.error(
          'Warning: SubMenuItem has a child which is not a MenuItem component',
        );
      }
    });
    return (
      <Transition in={menuOpen} timeout={300} animation="zoom-in-top">
        <ul className={subMenuClasses}>{childrenComponent}</ul>
      </Transition>
    );
  };
  return (
    <li key={index} className={classes} {...hoverEvents}>
      <div className="submenu-title" {...clickEvents}>
        {title}
        <Icon icon="angle-down" className="arrow-icon" />
      </div>
      {renderClidren()}
    </li>
  );
};

SubMenuItem.displayName = 'SubMenuItem';
export default SubMenuItem;
