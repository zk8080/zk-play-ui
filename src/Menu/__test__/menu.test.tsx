import React from 'react';
// import { render, RenderResult, fireEvent, wait } from '@testing-library/react';
import { render, shallow, mount, ReactWrapper } from 'enzyme';
import Menu, { MenuProps } from '../menu';
jest.mock('../../Icon', () => {
  return () => {
    return <i className="fa" />;
  };
});
jest.mock('react-transition-group', () => {
  return {
    CSSTransition: (props: any) => {
      return props.children;
    },
  };
});
const testProps: MenuProps = {
  defaultIndex: '0',
  onSelect: jest.fn(),
  className: 'test',
};
const testVerProps: MenuProps = {
  defaultIndex: '0',
  mode: 'vertical',
  defaultOpenSubMenus: ['4'],
};
const generateMenu = (props: MenuProps) => {
  return (
    <Menu {...props}>
      <Menu.Item>active</Menu.Item>
      <Menu.Item disabled>disabled</Menu.Item>
      <Menu.Item>xyz</Menu.Item>
      <Menu.SubMenuItem title="dropdown">
        <Menu.Item>drop1</Menu.Item>
      </Menu.SubMenuItem>
      <Menu.SubMenuItem title="opened">
        <Menu.Item>opened1</Menu.Item>
      </Menu.SubMenuItem>
    </Menu>
  );
};

let wrapper: ReactWrapper,
  wrapper2: ReactWrapper,
  menuElement: ReactWrapper,
  activeElement: ReactWrapper,
  disabledElement: ReactWrapper;
describe('test Menu and MenuItem component in default(horizontal) mode', () => {
  beforeEach(() => {
    wrapper = mount(generateMenu(testProps));
    menuElement = wrapper.find({ 'data-testid': 'test-menu' });
    activeElement = wrapper.find('.is-active');
    disabledElement = wrapper.find('.is-disabled');
  });
  it('should render correct Menu and MenuItem based on default props', () => {
    expect(menuElement).toMatchSnapshot();
    expect(menuElement.hasClass('viking-menu test')).toBeTruthy();
    expect(menuElement.find('li').length).toEqual(7);
    expect(activeElement.hasClass('menu-item is-active')).toBeTruthy();
    expect(disabledElement.hasClass('menu-item is-disabled')).toBeTruthy();
  });
  it('click items should change active and call the right callback', () => {
    const thirdItem = wrapper.find('.menu-item').at(2);
    thirdItem.simulate('click');
    expect(testProps.onSelect).toHaveBeenCalledWith('2');
    expect(
      wrapper.find('.menu-item').at(2).hasClass('menu-item is-active'),
    ).toBeTruthy();
    expect(
      wrapper.find('.menu-item').at(0).hasClass('menu-item is-active'),
    ).not.toBeTruthy();
    disabledElement.simulate('click');
    expect(disabledElement.hasClass('menu-item is-active')).not.toBeTruthy();
    expect(testProps.onSelect).not.toHaveBeenCalledWith('1');
  });
  it('should render the correct default menu with SubMenu', async () => {
    expect(wrapper.find('.submenu-item')).toHaveLength(2);
    expect(wrapper.find('.menu-item')).toHaveLength(7);
    expect(wrapper).toMatchSnapshot();
  });
});
describe('test Menu and MenuItem component in vertical mode', () => {
  beforeEach(() => {
    wrapper2 = mount(generateMenu(testVerProps));
  });
  it('should render vertical mode when mode is set to vertical', () => {
    const menuElement = wrapper2.find({ 'data-testid': 'test-menu' });
    expect(menuElement.hasClass('menu-vertical')).toBeTruthy();
  });
  it('should render correct dropdownItem and openedItem based on default props for vertical mode', () => {
    const dropDownItem = wrapper2.find({ title: 'dropdown' });
    expect(
      dropDownItem.find('.submenu-item').hasClass('is-opened'),
    ).not.toBeTruthy();
    const opendItem = wrapper2.find({ title: 'opened' });
    expect(opendItem.find('.submenu-item').hasClass('is-opened')).toBeTruthy();
  });
});
