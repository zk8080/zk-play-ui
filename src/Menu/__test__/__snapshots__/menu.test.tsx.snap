// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Menu and MenuItem component in default(horizontal) mode should render correct Menu and MenuItem based on default props 1`] = `ReactWrapper {}`;

exports[`test Menu and MenuItem component in default(horizontal) mode should render the correct default menu with SubMenu 1`] = `ReactWrapper {}`;
